Pegasus/include/Geometry.hpp

```diff
@@ -171,6 +171,468 @@ class Box : public SimpleShape
+ *          Object in given direction
+ *
+ *  Configuration Space Object (aka Minkowski Difference and Minkowski
+ *  Configuration Object) is a cartesian product of two sets of points, where
+ *  each element in one of the sets is multiplied by -1.
+ *
+ *  @param  box1        shape object
+ *  @param  box2        shape object
+ *  @param  direction   normalized search vector
+ *
+ *  @return point on the surface
+ */
+glm::dvec3 Support(Box const& box1, Box const& box2, glm::dvec3 direction);
+} // namespace cso
+
+namespace gjk
+{
+/** Simplex data container */
+struct Simplex
+{
+    //! Simplex vertices. Note that some vertices may be unused
+    std::array<glm::dvec3, 4> vertices;
+
+    //! Indicated number of used vertices
+    uint8_t size;
+};
+
+/**
+ *  @brief  Checks if simplex contains origin
+ *
+ *  @attention  Must only be called on a simplex of size in range [2; 4]
+ *
+ *  @param  simplex simplex data
+ *
+ *  @return @c true if simplex contains origin @c false otherwise
+ */
+bool SimplexContainsOrigin(Simplex const& simplex);
+
+/**
+ *  @brief  Calculates nearest simplex to the origin
+ *
+ *  Presumes that simplex vertices are stored in a way such that the latest
+ *  added vertex has index @c simplexSize - 1
+ *
+ *  Given simplex may be reduced down to size 1 as a result of this method.
+ *
+ *  @param[in,out]  simplex simplex data
+ *
+ *  @return new search direction
+ */
+glm::dvec3 NearestSimplex(Simplex& simplex);
+
+/**
+ *  @brief Checks if simplex contains origin
+ *
+ *  If simplex does not contain origin it is replaced by a new sub simplex
+ *  that is closest to the origin
+ *
+ *  @param[in,out]  simplex     current simplex
+ *  @param[in,out]  direction   current search direction
+ *
+ *  @return @c true if simplex contains origin, @c false otherwise
+ */
+bool DoSimplex(gjk::Simplex& simplex, glm::dvec3& direction);
+
+/**
+ *  @brief  Сalculates a tetrahedron from the CSO such that it contains the origin
+ *
+ *  If simplex contains origin then there is intersection between given shapes
+ *
+ *  @tparam ShapeA  any shape type for which gjk::Support is overloaded
+ *  @tparam ShapeB  any shape type for which gjk::Support is overloaded
+ *
+ *  @param[in,out]  simplex     initial simplex
+ *  @param[in]      aShape      reference to the shape object
+ *  @param[in]      bShape      reference to the shape object
+ *  @param[in]      direction   initial search direction vector of unit length
+ *
+ *  @return @c true if simplex contains origin, @c false otherwise
+ */
+template <typename ShapeA, typename ShapeB>
+bool CalculateSimplex(Simplex& simplex, ShapeA const& aShape, ShapeB const& bShape, glm::dvec3 direction)
+{
+    uint8_t counter = 0;
+    double previousPointDirectionProjection = std::numeric_limits<double>::min();
+
+    do
+    {
+        //Add new vertex to the simplex
+        simplex.vertices[simplex.size++] = cso::Support(aShape, bShape, direction);
```

Comment:
@dsodintsov This function holds the invariant, by this step the size of the simplex must be less or equal to 3.  It should not take a pre calculated simplex of size 4. The purpose of this function is to calculate one. So there is no way it would be 4.